translator/navjoy_translator.py

```diff
@@ -0,0 +1,347 @@
+    if index != None:
+        coordinates = map[index].get("coordinates", [None])[0]
+    else:
+        index = get_polygon_index(map)
+        if index != None:
+            polygon = map[index].get("coordinates", [None])[0]
+            coordinates = polygon_tools.polygon_to_polyline_center(polygon)
+        else:
+            logging.warning(
+                f"Invalid event with id = {id}. No LineString or Polygon found")
+            return False
+
+    if not coordinates:
+        logging.warning(
+            f"Invalid event with id = {obj.get('sys_gUid')}. No valid coordinates found")
+        return False
+
+    starttime_string = data.get('workStartDate')
+    endtime_string = data.get('workEndDate')
+    description = data.get('descriptionForProject')
+
+    required_fields = [street, starttime_string, description]
+    for field in required_fields:
+        if not field:
+            logging.warning(
+                f'''Invalid event with id = {id}. not all required fields are present. Required fields are: 
+                streetNameFrom, workStartDate, and descriptionForProject''')
+            return False
+
+    start_time = date_tools.parse_datetime_from_iso_string(starttime_string)
+    end_time = date_tools.parse_datetime_from_iso_string(endtime_string)
+    if not start_time:
+        logging.error(
+            f'Invalid incident with id = {id}. Unsupported start time format: {start_time}')
+        return False
+    elif endtime_string and not end_time:
+        logging.error(
+            f'Invalid incident with id = {id}. Unsupported end time format: {end_time}')
+        return False
+
+    return True
+
+
+# function to calculate vehicle impact
+def get_vehicle_impact(travelRestriction):
+    if not travelRestriction or type(travelRestriction) != str:
+        return None
+    travelRestriction = travelRestriction.lower()
+    vehicle_impact = 'all-lanes-open'
+    if 'lane closure' in travelRestriction.lower():
+        vehicle_impact = 'some-lanes-closed'
+    elif 'all lanes closed' in travelRestriction.lower():
+        vehicle_impact = 'all-lanes-closed'
+
+    return vehicle_impact
+
+
+def get_types_of_work(field):
+    if not field or type(field) != str:
+        return None
+    field = field.lower()
+    # valid_types_of_work = ['maintenance',
+    #                        'minor-road-defect-repair',
+    #                        'roadside-work',
+    #                        'overhead-work',
+    #                        'below-road-work',
+    #                        'barrier-work',
+    #                        'surface-work',
+    #                        'painting',
+    #                        'roadway-relocation',
+    #                        'roadway-creation']
+
+    if not field or type(field) != str:
+        return []
+
+    types_of_work = []
+
+    if 'restriping' in field or 'crack seal' in field:
+        types_of_work.append({'type_name': 'surface-work',
```

Comment:
We have quite a list of valid types above, but are only adding the surface-work here. Should we add a "//TODO" in here to mark additional types to be supported later?

Reply:
Yes, I will add a TODO. The complicating factor here is that none of the fields in navJoy are enums, they are all just human-entered notes. I am planning to improve this method in this sprint